File: src/fkt/java/util/Tracer.java
package fkt.java.util;

/**
Utility superclass that can issue trace messages.
 */
public abstract class Tracer implements Identified{
	public static int ids;
	public static class TracerTopped extends Tracer{
		private final String top;
		public TracerTopped(String top){
			super(top);
			this.top=top;
		}
		@Override
		protected void doTraceMsg(String msg){
			if(doTrace())super.doTraceMsg(msg);
		}
		protected boolean doTrace(){
			return true;
		}
	}
	private final String top;
	private Integer id=++ids;
	@Override
	public Object identity(){
		return id;
	}
	public static Tracer newTopped(final String top,final boolean live){
		return new TracerTopped(top);
	}
	public Tracer(String top){
		this.top=top;
	}
	public Tracer(){
		top=null;
	}
	final public void trace(String msg){
		doTraceMsg(msg);
	}
	final public void trace(String msg,Object o){
		if(o instanceof Object[])doTraceMsg(msg+newArrayText((Object[])o));		
		else doTraceMsg(msg+Debug.info(o));		
	}
	protected void doTraceMsg(String msg){
		Util.printOut((top!=null?(top+" #"+id):Debug.info(this))+" "+msg);
	}
	private String newArrayText(Object[]array){
		if(false)return Util.arrayPrintString(array);
		String lines=new String("[\n");
		for(Object o:array)
			lines+="  "+(true?Debug.info(o):o.toString())+"\n";
		lines+=("]");
		return lines;
	}
}

File: src/fkt/facets/util/Tracer.kt
package fkt.facets.util
abstract class Tracer(protected val top:String?=null): Identified {
  private val id = ++ids
  private class Topped(top: String):Tracer(top) {
    override fun doTraceMsg(msg:String) {
      if (doTrace()) println("$top: $msg")
    }
    private fun doTrace():Boolean {
      return true
    }
  }
  override val identity=id
  fun trace(msg:String) {
    doTraceMsg(msg)
  }
  fun trace(msg:String, o:Any) {
      val trace = msg + when (o) {
        is Collection<*> -> newArrayText(o.toTypedArray())
        is Array<*> -> newArrayText(o)
        else -> Debug.info(o)
      }
      doTraceMsg(trace)
  }
  protected open fun doTraceMsg(msg:String) {
    println((if (top != null) ("$top #$id") else Debug.info(this)) + " " + msg)
  }
  private fun newArrayText(array:Array<*>):String {
    if (false) return Util.arrayPrintString(array)
    var lines = "[\n"
    for (o in array)
      lines += " " + (if (true) Debug.info(o) else o.toString()) + "\n"
    lines += ("]")
    return lines
  }
  companion object {
    var ids:Int = 0
    fun newTopped(top:String):Tracer=Topped(top)
  }
}